src/js/utils/linked-list.js

```diff
@@ -1,83 +1,84 @@
 export default class LinkedList {
-  constructor(options) {
+  constructor(options={adoptItem: () => {}, freeItem: () => {}}) {
     this.head = null;
     this.tail = null;
-    if (options) {
-      let {adoptItem, freeItem} = options;
-      this.adoptItem = adoptItem;
-      this.freeItem = freeItem;
-    }
+    this.adoptItem = options.adoptItem;
+    this.freeItem = options.freeItem;
   }
   prepend(item) {
-    this.insertBefore(item, this.head);
+    this.adoptItem(item);
+
+    if (this.head) {
+      this.insertBefore(item, this.head);
+    } else {
+      this.head = item;
+      this.tail = item;
+      item.prev = null;
+      item.next = null;
+    }
   }
   append(item) {
-    this.insertBefore(item, null);
+    this.adoptItem(item);
+
+    if (this.tail) {
+      this.insertAfter(item, this.tail);
+    } else {
+      this.prepend(item);
+    }
   }
-  insertAfter(item, prevItem) {
-    let nextItem = null;
-    if (prevItem) {
-      nextItem = prevItem.next;
+  insertAfter(newItem, oldItem) {
+    this.adoptItem(newItem);
+
+    if (!oldItem) {
+      this.append(newItem);
+      return;
     }
-    this.insertBefore(item, nextItem);
+
+    newItem.prev = oldItem;
+    newItem.next = oldItem.next;
+
+    if (oldItem.next) {
+      oldItem.next.prev = newItem;
+    } else {
+      this.tail = newItem;
+    }
+
+    oldItem.next = newItem;
   }
-  insertBefore(item, nextItem) {
-    this.remove(item);
-    if (this.adoptItem) {
-      this.adoptItem(item);
+  insertBefore(newItem, oldItem) {
+    this.adoptItem(newItem);
+
+    if (!oldItem) {
+      this.prepend(newItem);
```

Comment:
Because `prepend` also adopts, we would end up calling `adoptItem` twice. This was the motivation for using one `insertBefore` method, it also makes it trivial to know that edge cases for any particular method would be present in all methods (simpler to debug IMO, at the expense of being harder to read).